eggplant/market/models/cart.py

```diff
@@ -53,35 +60,49 @@ def add_to_items(self, product=None, quantity=1, delivery_date=None):
 
     def remove_from_items(self, product=None, quantity=1, delivery_date=None):
         current = self.items.filter(product=product,
-                                    delivery_date=delivery_date)
-        if current.exists():
-            if current[0].quantity - quantity > 1:
-                current.update(quantity=models.F('quantity') - quantity)
+                                    delivery_date=delivery_date).first()
+        if current:
+            if current.quantity - quantity > 1:
+                current.quantity -= quantity
+                current.save()
             else:
                 self.items.filter(basket=self, product=product, delivery_date=delivery_date).delete()
 
     def get_total_amount(self):
-        # TODO: This should return a Money type instead of decimal
-        # TODO: When calculating this, we should fail when currencies differ
-        total = Decimal('0')
-        for item in self.items.all():
-            total += item.quantity * item.product.price.amount
-        return total
+
+        # FIXME this check should be removed, it is better to check for the currenices when an item is added
+        # sanity check for the currencies
+
+        if 1 != self.items.all().values('product__price_currency').annotate(total=Count('product__price_currency')).count():
+            raise ValidationError('The products have different currencies')
+
+        total = self.items.all().aggregate(total=Sum(F('product__price') * F('quantity')), cur=Max('product__price_currency'))
+        return Money(total['total'], total['cur'])
```

Comment:
This doesn't work for me, I get the following:

```
site-packages/django/db/models/expressions.py in _resolve_output_field(self)
    268                     if source is not None and not isinstance(self._output_field, source.__class__):
    269                         raise FieldError(
--> 270                             "Expression contains mixed types. You must set output_field")
    271 
    272     def convert_value(self, value, expression, connection, context):

FieldError: Expression contains mixed types. You must set output_field
```

I think it has to do with the `F('product__price') * F('quantity')` expression.

I'm guessing it somehow works for you? Are you running Python 3.4 or 3.5?